File: LC_Medium/kth_smallest_element_in_BST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):    
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """

        if not root.left and not root.right:
            return root.val

        output = []
        self.kthSmallestRec(root, k, output)

        return output[len(output) - 1]


    def kthSmallestRec(self, root, k, output):
        """
        :type root: TreeNode
        :type k: int
        """

        if root.left:
            self.kthSmallestRec(root.left, k, output)

        if len(output) is k:
            return

        output.append(root.val)            
        
        if root.right:
            self.kthSmallestRec(root.right, k, output)

File: LC_Medium/kth_smallest_element_in_BST.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    TreeNode found = null;
    public int KthSmallest(TreeNode root, int k) 
    {
        if (root.left == null && root.right == null)
        {
            return root.val;
        }

        KthSmallestByRef(root, ref k);

        return found.val;
    }

    public void KthSmallestByRef(TreeNode root, ref int k)
    {
        if (root == null)
        {
            return;
        }
        
        KthSmallestByRef(root.left, ref k);
        k--;
        if (k == 0)
        {
            found = root;
            return;
        }
        KthSmallestByRef(root.right, ref k);
    }
}